Fix Memory Leak in `EVP_MD_CTX_free` by Removing EVP_MD_CTX_FLAG_REUSE

erbsland-dev opened the pull request:
Alternative fix for #6343: Address a memory leak in the `EVP_MD_CTX_free` function caused by the `EVP_MD_CTX_FLAG_REUSE` flag, which was likely introduced to optimize the copy performance of `EVP_MD_CTX`. Eliminate the internal flag `EVP_MD_CTX_FLAG_REUSE` and its usage. Introduce the `evp_md_ctx_take_data` helper function to enhance code readability. Refactor `EVP_MD_CTX_copy_ex`: Update the legacy implementation by using the new helper function and reorganize the logic to group related code segments for better readability.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [ ] documentation is added or updated
- [ ] tests are added or updated

t8m commented:
This is IMO wrong and we should not try to fix this misfeature. Also the CI failures are relevant.

t8m commented:
Also you cannot remove public #define - you could probably still remove its internal usage but I do not see a need for this.

erbsland-dev commented:
@t8m 
> This is IMO wrong and we should not try to fix this misfeature. Also the CI failures are relevant.
> Also you cannot remove public #define - you could probably still remove its internal usage but I do not see a need for this.

I understand and agree with your opinion. I also prefer the simpler implementation where the flag is cleared at a suitable place. However, I believe it’s sometimes helpful to visualize alternatives, and this PR aims to provide such a visualization.

I’ve updated the code to restore the `#define` to the API and fixed the memory leak I introduced, replaced it by a crash in the GOST algorithms. It is interesting to see the CI errors that are raised by the changes.
